packages/flutter/lib/src/material/material_state_mixin.dart

```diff
@@ -0,0 +1,130 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'package:flutter/material.dart';
+
+import '../../foundation.dart';
+
+/// Provides a flexible callback and consistent API for [State] objects to track
+/// which [MaterialState] values are in effect for their portion of the user
+/// interface.
+///
+/// This mixin does nothing by merely applying it to a [State] class, but is
+/// helpful when writing `build` methods that include child [InkWell],
+/// [GestureDetector], [MouseRegion], or [Focus] widgets. Instead of manually
+/// creating handlers for each type of user interaction, such [State] classes can
+/// instead provide a `ValueChanged<bool>` function and allow [MaterialStateMixin]
+/// to manage the set of active [MaterialState]s, and the calling of `setState()`
+/// as necessary.
+///
+/// Use of this mixin changes the implementation below, in Example 1, into the
+/// ensuing, much shorter implementation in Example 2. The two code samples are
+/// functionally equivalent.
+///
+/// Example 1: Tracking MaterialState values without MaterialStateMixin.
```

Comment:
these examples should be wrapped in `{@tool ...} ... {@endtool}` so they get analyzed

Reply:
Also: Maybe we should just show the "with MaterialStateMixin" example in the docs to make sure people don't copy the wrong thing.